File: snappii/app/src/main/java/com/home/snappii/di/NetworkModule.java
package com.home.snappii.di;

import android.app.Application;

import com.home.snappii.AppApplication;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@Module
public class NetworkModule {


    private Application application;

    public NetworkModule(AppApplication application) {
        this.application = application;
    }

    @Singleton
    @Provides
    @Named("media")
    String providesMediaUrl() {
        return "https://randomuser.me/";
    }

    @Provides
    @Named("network")
    Scheduler providesSubscribeScheduler() {
        return Schedulers.io();
    }

    @Provides
    @Named("ui")
    Scheduler providesObserveScheduler() {
        return AndroidSchedulers.mainThread();
    }

}

File: Trucks/app/src/test/java/ru/rsprm/di/NetworkModule.kt
package ru.rsprm.di

import android.content.Context
import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import io.reactivex.schedulers.Schedulers
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import ru.rsprm.network.Api
import ru.rsprm.network.ConnectionInterceptor
import ru.rsprm.network.NetworkUtils
import javax.inject.Named
import javax.inject.Singleton


@Module
class NetworkModule(val context: Context, val url: String) {

    @Singleton
    @Named(NETWORK)
    @Provides
    fun provideNetworkScheduler(): Scheduler {
        return Schedulers.trampoline()
    }

    @Singleton
    @Named(UI)
    @Provides
    fun provideUIScheduler(): Scheduler {
        return Schedulers.trampoline()
    }

    @Singleton
    @Provides
    fun provideNetworkUtils(): NetworkUtils {
        return NetworkUtils()
    }

    @Singleton
    @Provides
    fun provideOkHttpClient(networkUtils: NetworkUtils): OkHttpClient {
        val logging = HttpLoggingInterceptor()
        logging.setLevel(HttpLoggingInterceptor.Level.BODY)

        val httpClient = OkHttpClient
            .Builder()
            .addInterceptor(ConnectionInterceptor(context, networkUtils))
            .addInterceptor(logging)
            .build()

        return httpClient
    }

    @Singleton
    @Provides
    fun provideApiService(httpClient: OkHttpClient): Api {
        val retrofit = Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .client(httpClient)
            .baseUrl(url)
            .build()

        return retrofit.create(Api::class.java)
    }
}
